src/basic/fd-util.h

```diff
@@ -120,3 +124,21 @@ static inline char *format_proc_fd_path(char buf[static PROC_FD_PATH_MAX], int f
 
 #define FORMAT_PROC_FD_PATH(fd) \
         format_proc_fd_path((char[PROC_FD_PATH_MAX]) {}, (fd))
+
+/* Determining a file descriptor's path via /proc is not reliable. Only use this function to provide better
+ * log messages when doing fd based path handling! */
+static inline const char *FORMAT_FD_PATH(int fd, char **buf) {
+        assert(buf);
+
+        if (fd == AT_FDCWD)
+                return "";
+
+        *buf = mfree(*buf);
```

Comment:
That's unusual. Why this?

Reply:
To make it easier to use this function. Ideally we wouldn't have to pass the buffer in at all, but see discussion above for why that's not possible. If we have to pass in a buffer, let's at least allow callers to pass the same buffer everywhere and automatically free it for them so they don't have to do it manually themselves every time.